server/build.gradle

```diff
@@ -151,8 +151,19 @@ if (BuildParams.isSnapshotBuild() == false) {
     }
 }
 
+tasks.register('testNetworkTraceNotEnabled', Test) {
```

Comment:
I'm not a huge fan of introducing another test task here. What exaclty does the `es.insecure_network_trace_enabled` system property do?

Reply:
I'm absolutely on the same page with you, I just don't see much of an alternative. Also, I admit, having `test` depend on `testNetworkTraceNotEnabled` is broken. If you run `:server:test --tests XYZ` the dependency would still run ... 

`es.insecure_network_trace_enabled` sets this flag, which gates the behavior added in this PR. Goal of the PR is to prevent setting dangerous loggers that might leak sensitive information. However, in order to investigate test failures, we are enabling this flag for most tests including `:server:test`


```java
public class NetworkTraceFlag {

    private NetworkTraceFlag() {
        // no instances;
    }

    public static final String PROPERTY_NAME = "es.insecure_network_trace_enabled";

    public static final boolean TRACE_ENABLED;

    static {
        final var propertyValue = System.getProperty(PROPERTY_NAME);
        if (propertyValue == null) {
            TRACE_ENABLED = false;
        } else if ("true".equals(propertyValue)) {
            TRACE_ENABLED = true;
        } else {
            throw new IllegalArgumentException(
                Strings.format("system property [%s] may only be set to [true], but was [%s]", PROPERTY_NAME, propertyValue)
            );
        }
    }
}
```

Reply:
Hmm, is there anyway we can finagle this in a way that would allow updating this property inside specific tests? Initializing statics makes this tricky but perhaps some way to "reinitialize" this flag dynamically?

We definitely don't want test tasks depending on other test tasks. At _worst_ we'd have `check` depend on this additional task.

Reply:
> Hmm, is there anyway we can finagle this in a way that would allow updating this property inside specific tests? Initializing statics makes this tricky but perhaps some way to "reinitialize" this flag dynamically?

IMHO that would pretty much defeat the purpose of that flag.

An alternative to a separate test task (using a different configuration) might be to extend the test runner to support forking tests into a new JVM. Though, while convenient, something like would probably be quickly adopted all around with a terrible impact on test run performance :/

Reply:
I wonder if we're overthinking this. AIUI we are trying to run the entire test suite twice, to ensure coverage of both states of this flag. That seems unnecessary to me, I think we only really need to run a few specific tests with tracing on (or off) and the rest of the test suite needn't care either way. We could make a dedicated suite in `/qa` for the otherwise-untested state. Or we could randomize it in `:server:test` and have two dedicated suites in `/qa` just for the tests that need it switched on or off.

Reply:
Note that we have some other `:qa` suites that are similarly tightly-focussed, e.g. `:qa:unconfigured-node-name` and `:qa:custom-rest-controller`.

Reply:
+1 to having an explicit QA project for this.

Reply:
@rjernst I think this is a good compromise: In https://github.com/elastic/elasticsearch/pull/105020/commits/7f57feb5e739c73fd8810a038454d285c38360b2 I've moved the tests to `:qa:restricted-loggers`. But I am also keeping similar tests in `:server` that are using an internal API allowing to explicitly pass restricted loggers. This should address your concern to keep tests close by the code. And at the same time we don't have to compromise on the static / final initialization of `TRACE_ENABLED` / `RESTRICTED_LOGGERS`.